taskmgr/queueservice.go

```diff
@@ -0,0 +1,387 @@
+package taskmgr
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	logrus "github.com/Sirupsen/logrus"
+	"github.com/taskcluster/httpbackoff"
+	tcqueue "github.com/taskcluster/taskcluster-client-go/queue"
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+)
+
+type (
+	// Used for modelling the xml we get back from Azure
+	QueueMessagesList struct {
+		XMLName       xml.Name       `xml:"QueueMessagesList"`
+		QueueMessages []QueueMessage `xml:"QueueMessage"`
+	}
+
+	// Used for modelling the xml we get back from Azure
+	QueueMessage struct {
+		XMLName         xml.Name        `xml:"QueueMessage"`
+		MessageId       string          `xml:"MessageId"`
+		InsertionTime   azureTimeFormat `xml:"InsertionTime"`
+		ExpirationTime  azureTimeFormat `xml:"ExpirationTime"`
+		DequeueCount    uint            `xml:"DequeueCount"`
+		PopReceipt      string          `xml:"PopReceipt"`
+		TimeNextVisible azureTimeFormat `xml:"TimeNextVisible"`
+		MessageText     string          `xml:"MessageText"`
+	}
+
+	// Custom time format to enable unmarshalling of azure xml directly into go
+	// object with native go time.Time implementation under-the-hood
+	azureTimeFormat struct {
+		time.Time
+	}
+
+	// TaskId and RunId are taken from the json encoding of
+	// QueueMessage.MessageId that we get back from Azure
+	TaskRun struct {
+		TaskId              string                         `json:"taskId"`
+		RunId               uint                           `json:"runId"`
+		SignedDeleteUrl     string                         `json:"-"`
+		TaskClaimResponse   tcqueue.TaskClaimResponse      `json:"-"`
+		TaskReclaimResponse tcqueue.TaskReclaimResponse    `json:"-"`
+		Definition          tcqueue.TaskDefinitionResponse `json:"-"`
+		reclaimTimer        *time.Timer
+	}
+
+	TaskQueue struct {
+		SignedDeleteUrl string `json:"signedDeleteUrl"`
+		SignedPollUrl   string `json:"signedPollUrl"`
+	}
+
+	TaskRuns []*TaskRun
+
+	queueService struct {
+		mu               sync.Mutex
+		queues           []TaskQueue
+		Expires          tcclient.Time
+		ExpirationOffset int
+		client           *tcqueue.Queue
+		ProvisionerId    string
+		WorkerType       string
+		WorkerId         string
+		WorkerGroup      string
+		Log              *logrus.Entry
+	}
+)
+
+// Given a number of tasks needed, the Azure task queues will be polled in order
+// of priority until either there are no more tasks to claim, or the given number of
+// tasks has been fulfilled.
+func (q queueService) ClaimWork(ntasks int) []*TaskRun {
```

Comment:
Not saying we have to refactor this right now... But I think the smart interface for this is:
```
taskpoller.New(config, func() time.Duration { return pollDelay }, func(run *TaskRun) { /* start new task */ });
taskpoller.Start() // start polling when tasks are needed
taskpoller.RequestTasks(n int) // Request n tasks to be claimed and /* start new task */ be called in new go routine
// After finishing a task we shall call taskpoller.RequestTasks(1) 
// Initially we call taskpoller.RequestTasks(capacity)
taskpoller.Stop() // Stop polling
```

This way we can start a task as soon as we get it... And we can increasing the maximum number of tasks we're willing to claim between two requests, say after trying the highest queue and finding that it had nothing...

That said, I think this interface isn't too bad either :)

The first func argument to New is all about letting something else decide how long to sleep between polling everything again... I guess in the current interface that isn't controlled here... Which might be okay too.. I need to read me.. :)